src/tracks/ui/TextEditHelper.cpp

```diff
@@ -0,0 +1,666 @@
+    {
+        if (mBBox.Contains(event.GetPosition()))
+        {
+            if (event.LeftDown()) 
+            {
+                mRightDragging = false;
+                auto position = FindCursorIndex(event.GetPosition());
+                auto initial = mInitialCursorPos;
+                if (event.ShiftDown()) {
+#ifdef __WXMAC__
+                    // Set the drag anchor at the end of the previous selection
+                    // that is farther from the NEW drag end
+                    const auto current = mCurrentCursorPos;
+                    if (abs(position - current) > abs(position - initial))
+                        initial = current;
+#else
+                    // initial position remains as before
+#endif
+                }
+                else
+                    initial = position;
+
+                mInitialCursorPos = initial;
+                mCurrentCursorPos = position;
+            }
+            else
+            {
+                if (mInitialCursorPos == mCurrentCursorPos)
+                {
+                    auto position = FindCursorIndex(event.GetPosition());
+                    mInitialCursorPos = mCurrentCursorPos = position;
+                }
+                // Actually this might be right or middle down
+                mRightDragging = true;
+            }
+
+#if defined(__WXGTK__) && (HAVE_GTK)
+            if (evt.MiddleDown()) {
+                // Paste text, making a NEW label if none is selected.
+                wxTheClipboard->UsePrimarySelection(true);
+                view.PasteSelectedText(project, newSel.t0(), newSel.t1());
+                wxTheClipboard->UsePrimarySelection(false);
+}
+#endif
+            return true;
+        }
+    }
+    return false;
+}
+
+bool TextEditHelper::OnDrag(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+bool TextEditHelper::OnRelease(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+void TextEditHelper::Draw(wxDC& dc, const wxRect& rect)
+{
+    mBBox = rect;
+    dc.SetFont(mFont);
+
+    const auto cursorHeight = dc.GetFontMetrics().height;
+
+    wxRect prevClippingRegion;
+    auto restoreClippingRegion = dc.GetClippingBox(prevClippingRegion);
+    dc.SetClippingRegion(rect);
+
+    
+    auto rtl = wxTheApp->GetLayoutDirection() == wxLayout_RightToLeft;
+
+    auto curPosX = 0;
+    auto maxOffset = static_cast<int>(mText.Length());
+    mOffset = std::clamp(mOffset, 0, maxOffset);
+    {
+        auto leftBound = rect.GetLeft();
+        auto rightBound = rect.GetRight() + 1;
+        GetCharPositionX(mCurrentCursorPos, &curPosX);
+
+        if ((!rtl && curPosX >= rightBound) || (rtl && curPosX < leftBound))
+        {
+            while (mOffset < maxOffset)
+            {
+                GetCharPositionX(mCurrentCursorPos, &curPosX);
+                if (curPosX < rightBound && curPosX >= leftBound)
```

Comment:
Is this condition correct in case of right-to-left layout?

Reply:
Yep, don't work as expected with RTL layout and LTR language